tests: Initial unit tests

jcvenegas opened the pull request:
- Add initial utils.go
 - Add .ci/go-test.go for CI
- Add .ci/go-lint to run gometalinter
- Add make check target to run go-test.sh and go-lint.sh

Fixes: #21

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
tests: Initial unit tests

 - Add initial utils.go
 - Add .ci/go-test.go for CI

Fixes: #21

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas commented on Makefile:
```diff
@@ -53,6 +53,8 @@ $(GENERATED_FILES): %: %.in Makefile
 		-e 's|[@]ccagent[@]|$(TARGET)|g' \
 		"$<" > "$@"
+check:
+	./.ci/go-test.sh
```
note to myself: This need to be a relative path to this makefile  so we call $(MKDIR)/.ci/go-test.sh

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+	"testing"
+)
+
+func Test_fileCopy(t *testing.T) {
```
Please follow the Golang syntax by naming this function:
```go
func TestFileCopy(t *testing.T)
```

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+		args    args
+		wantErr bool
+	}{
+		{"Source/destination path are empty",
```
s/path/paths/

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+	type args struct {
+		srcPath string
+		dstPath string
+	}
```
Please add a blank line here.

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+			args{srcPath: "./f/f/false", dstPath: "./d"}, true},
+		{"Destination path does not exist",
+			args{srcPath: tmpfile.Name(), dstPath: tmpfile.Name() + "/non-existing/dest"}, true},
+		{"Destination path does not exist",
```
You should differentiate the last two tests with different descriptions.

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+			if (err != nil) != tt.wantErr {
+				t.Errorf("fileCopy() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if err != nil {
```
You can remove this `err` check if you call into `os.RemoveAll()` instead of `os.Remove()`. Indeed, the `os.RemoveAll()` won't generate any issue even if the file does not exist.

sboeuf commented on utils_test.go:
```diff
@@ -0,0 +1,69 @@
+		t.Run(tt.name, func(t *testing.T) {
+			err := fileCopy(tt.args.srcPath, tt.args.dstPath)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("fileCopy() error = %v, wantErr %v", err, tt.wantErr)
```
Why `t.Errorf()` instead of `t.Fatalf()` here ?

jcvenegas replied:
It is to allow to allow to do extra checks/tests/cleanup in the current subtests in the future. We know that it is an error but is not something fatal. I based the implementation in https://blog.golang.org/subtests

sboeuf reviewed (changes requested):
Few comments on your PR.
One more thing, I would like to see the gometalinter run on every PR to make sure about things like `go vet`, `go fmt`, `gocyclo`, ....
Could you add this to the PR ? You can take virtcontainers as an example if you want.

jcvenegas committed:
tests: Add syscall.go unit tests

This patch cherrypicks  test for syscall.go from

github.com/containers/virtcontainers/syscall_test.go

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas committed:
Makefile: Add make check target

Add check target to run go unit tets

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

jcvenegas commented:
@grahamwhaley  @sboeuf seems that semaphore caught and in issue. I dont think is related with this PR I am creating an issue (#69) before update this PR

sboeuf commented on Makefile:
```diff
@@ -53,6 +54,9 @@ $(GENERATED_FILES): %: %.in Makefile
 		"$<" > "$@"
+check:
+	$(MKDIR)/.ci/go-test.sh
+	$(MKDIR)/.ci/go-lint.sh
```
We want the linter test first because we don't want to test the code if the syntax is wrong.

grahamwhaley replied:
+1 I think. It will cost us little, but it is the right thing to do :-)

sboeuf commented:
@jcvenegas please re-organize your commits so that go-lint.sh is introduced before the Makefile change calling into this script.

jcvenegas committed:
ci/test: Add gometalinter

This patch adds gometalinter and will run with
make check.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>

sboeuf commented:
LGTM

[![Approved with PullApprove](https://img.shields.io/badge/approvers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/64/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent) [![Approved with PullApprove](https://img.shields.io/badge/reviewers-approved-brightgreen.svg)](https://pullapprove.com/clearcontainers/agent/pull-request/64/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=clearcontainers/agent)

jcvenegas committed:
tests: Fix linter issues

Fix linter issues need run gometalinter in our CI.

Signed-off-by: Jose Carlos Venegas Munoz <jose.carlos.venegas.munoz@intel.com>
